File: internal/registry/registry.go
package registry

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
	"math"
	"sort"
	"sync"
	"time"

	"github.com/Rakshita-0023/conduit/internal/config"
	"github.com/Rakshita-0023/conduit/internal/policy"
	"github.com/modelcontextprotocol/go-sdk/mcp"
)

type Catalog struct {
	ServerID string
	Tools    []*mcp.Tool
}

// Route is the complete destination information for an advertised tool. It is
// deliberately not derived from PublicName at lookup time.
type Route struct {
	PublicName         string
	ServerID           string
	DownstreamToolName string
}

type AggregateState string

const (
	StateStarting    AggregateState = "starting"
	StateReady       AggregateState = "ready"
	StateUnavailable AggregateState = "unavailable"
	StateCollision   AggregateState = "aggregate_collision"
	StateOverLimit   AggregateState = "aggregate_over_limit"
)

// Snapshot is immutable once returned by Registry. Tools and routes are built
// before publication and are never subsequently modified by Registry.
type Snapshot struct {
	Generation uint64
	CreatedAt  time.Time
	State      AggregateState
	ToolCount  int

	tools      []*mcp.Tool
	routes     map[string]Route
	execution  map[string]executionRoute
	resultJSON json.RawMessage
	resultSize int64
}

func (s *Snapshot) Resolve(name string) (Route, bool) {
	route, ok := s.routes[name]
	return route, ok
}

func (s *Snapshot) ResultJSON() []byte { return append([]byte(nil), s.resultJSON...) }
func (s *Snapshot) ResultSize() int64  { return s.resultSize }

type Registry struct {
	mu       sync.RWMutex
	catalogs map[string]Catalog
	policy   policy.Compiled
	limits   config.Limits
	impl     mcp.Implementation
	active   *Snapshot
}

type executionRoute struct {
	route Route
	tool  *mcp.Tool
}

type PreparedRoute struct {
	Generation   uint64
	Route        Route
	Tool         *mcp.Tool
	PolicyDigest string
}

var (
	ErrRouteChanged = errors.New("route changed")
	ErrRouteDenied  = errors.New("route denied")
	ErrRouteMissing = errors.New("route missing")
)

var errAggregateResponseOverLimit = errors.New("aggregate response exceeds byte limit")

func New(limits config.Limits, compiled policy.Compiled, impl mcp.Implementation) *Registry {
	return &Registry{
		catalogs: map[string]Catalog{},
		policy:   compiled,
		limits:   limits,
		impl:     impl,
		active:   &Snapshot{State: StateStarting, CreatedAt: time.Now().UTC()},
	}
}

// Publish takes ownership of one complete validated catalog, then atomically
// publishes a complete aggregate state derived from all current catalogs.
func (r *Registry) Publish(c Catalog) (*Snapshot, error) {
	owned, err := cloneCatalog(c)
	if err != nil {
		return nil, err
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	r.catalogs[owned.ServerID] = owned
	r.rebuildLocked()
	return r.active, nil
}

// Remove discards a catalog that is no longer currently valid. No stale tools
// from that downstream can survive the resulting rebuild.
func (r *Registry) Remove(serverID string) *Snapshot {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.catalogs, serverID)
	r.rebuildLocked()
	return r.active
}

func (r *Registry) Snapshot() *Snapshot {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.active
}

func (r *Registry) Count(id string) int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.catalogs[id].Tools)
}

// PrepareExecution returns a caller-owned clone of the original downstream
// tool definition. It intentionally does not expose execution data through
// Snapshot, whose public routes remain discovery-filtered.
func (r *Registry) PrepareExecution(name string) (PreparedRoute, error) {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if r.active.State != StateReady {
		return PreparedRoute{}, ErrRouteMissing
	}
	entry, ok := r.active.execution[name]
	if !ok {
		return PreparedRoute{}, ErrRouteMissing
	}
	tool, err := cloneTool(entry.tool)
	if err != nil {
		return PreparedRoute{}, fmt.Errorf("clone execution tool: %w", err)
	}
	return PreparedRoute{Generation: r.active.Generation, Route: entry.route, Tool: tool, PolicyDigest: r.policy.Digest()}, nil
}

// CommitAuthorization creates the authorization linearization point. The
// callback must only durably write the audit record; it must not call Registry,
// health, or refresh code while this read lock prevents publication.
func (r *Registry) CommitAuthorization(prepared PreparedRoute, commit func() error) error {
	r.mu.RLock()
	defer r.mu.RUnlock()
	if r.active.State != StateReady || r.active.Generation != prepared.Generation {
		return ErrRouteChanged
	}
	entry, ok := r.active.execution[prepared.Route.PublicName]
	if !ok {
		return ErrRouteMissing
	}
	if entry.route != prepared.Route {
		return ErrRouteChanged
	}
	if !r.policy.Allowed(prepared.Route.PublicName) {
		return ErrRouteDenied
	}
	return commit()
}

func (r *Registry) rebuildLocked() {
	next := &Snapshot{Generation: r.active.Generation + 1, CreatedAt: time.Now().UTC(), State: StateUnavailable}
	if len(r.catalogs) == 0 {
		r.active = next
		return
	}

	type candidate struct {
		tool       *mcp.Tool
		downstream *mcp.Tool
		route      Route
	}
	byName := make(map[string]candidate)
	for _, catalog := range r.catalogs {
		for _, downstream := range catalog.Tools {
			if downstream == nil || downstream.Name == "" {
				r.active = next
				return
			}
			publicName := catalog.ServerID + "." + downstream.Name
			if _, exists := byName[publicName]; exists {
				next.State = StateCollision
				r.active = next
				return
			}
			publicTool := *downstream
			publicTool.Name = publicName
			byName[publicName] = candidate{tool: &publicTool, downstream: downstream, route: Route{PublicName: publicName, ServerID: catalog.ServerID, DownstreamToolName: downstream.Name}}
		}
	}

	next.execution = make(map[string]executionRoute, len(byName))
	for name, entry := range byName {
		next.execution[name] = executionRoute{route: entry.route, tool: entry.downstream}
	}
	names := make([]string, 0, len(byName))
	for name := range byName {
		if r.policy.Allowed(name) {
			names = append(names, name)
		}
	}
	sort.Strings(names)
	if len(names) > r.limits.MaxAggregateTools {
		next.State = StateOverLimit
		r.active = next
		return
	}
	next.tools = make([]*mcp.Tool, 0, len(names))
	next.routes = make(map[string]Route, len(names))
	for _, name := range names {
		entry := byName[name]
		next.tools = append(next.tools, entry.tool)
		next.routes[name] = entry.route
	}
	encoded, err := encodeListResult(next.tools, r.impl, r.limits.MaxAggregateResponseBytes)
	if err != nil {
		if errors.Is(err, errAggregateResponseOverLimit) {
			next.State = StateOverLimit
		}
		next.tools = nil
		next.routes = nil
		next.execution = nil
		r.active = next
		return
	}
	next.State = StateReady
	next.ToolCount = len(next.tools)
	next.resultJSON = encoded
	next.resultSize = int64(len(encoded))
	r.active = next
}

// encodeListResult is the sole client-facing tools/list result constructor.
// Registry uses its bytes for the aggregate limit and ingress writes these
// exact bytes as the JSON-RPC result value.
func encodeListResult(tools []*mcp.Tool, impl mcp.Implementation, limit int64) ([]byte, error) {
	var result boundedBuffer
	result.limit = limit
	if err := result.writeString(`{"resultType":"complete","_meta":{"io.modelcontextprotocol/serverInfo":`); err != nil {
		return nil, err
	}
	serverInfo, err := json.Marshal(impl)
	if err != nil {
		return nil, fmt.Errorf("encode aggregate server information: %w", err)
	}
	if err := result.Write(serverInfo); err != nil {
		return nil, err
	}
	if err := result.writeString(`},"ttlMs":0,"cacheScope":"public","tools":[`); err != nil {
		return nil, err
	}
	for i, tool := range tools {
		if i > 0 {
			if err := result.writeString(","); err != nil {
				return nil, err
			}
		}
		encodedTool, err := json.Marshal(tool)
		if err != nil {
			return nil, fmt.Errorf("encode aggregate tool: %w", err)
		}
		if err := result.Write(encodedTool); err != nil {
			return nil, err
		}
	}
	if err := result.writeString(`]}`); err != nil {
		return nil, err
	}
	return result.Bytes(), nil
}

// boundedBuffer retains at most limit+1 bytes. The extra byte distinguishes
// an exact limit from overflow without retaining a complete oversized result.
type boundedBuffer struct {
	bytes.Buffer
	limit int64
}

func (b *boundedBuffer) Write(p []byte) error {
	maximumRetained := b.limit
	if maximumRetained < math.MaxInt64 {
		maximumRetained++
	}
	remaining := maximumRetained - int64(b.Len())
	if remaining <= 0 {
		return errAggregateResponseOverLimit
	}
	if int64(len(p)) > remaining {
		_, _ = b.Buffer.Write(p[:remaining])
		return errAggregateResponseOverLimit
	}
	_, _ = b.Buffer.Write(p)
	if int64(b.Len()) > b.limit {
		return errAggregateResponseOverLimit
	}
	return nil
}

func (b *boundedBuffer) writeString(s string) error { return b.Write([]byte(s)) }

func cloneCatalog(c Catalog) (Catalog, error) {
	owned := Catalog{ServerID: c.ServerID, Tools: make([]*mcp.Tool, len(c.Tools))}
	for i, tool := range c.Tools {
		if tool == nil {
			continue
		}
		copy, err := cloneTool(tool)
		if err != nil {
			return Catalog{}, fmt.Errorf("clone catalog tool: %w", err)
		}
		owned.Tools[i] = copy
	}
	return owned, nil
}

func cloneTool(tool *mcp.Tool) (*mcp.Tool, error) {
	if tool == nil {
		return nil, nil
	}
	data, err := json.Marshal(tool)
	if err != nil {
		return nil, err
	}
	copy := new(mcp.Tool)
	if err := json.Unmarshal(data, copy); err != nil {
		return nil, err
	}
	return copy, nil
}

File: src/conduit/registry.py
"""Immutable aggregate generations and explicit public tool routes."""

from __future__ import annotations

import asyncio
import json
from dataclasses import dataclass
from types import MappingProxyType
from typing import Any, Awaitable, Callable, Mapping

from .config import LimitsConfig
from .policy import Policy

STARTING = "starting"
READY = "ready"
UNAVAILABLE = "unavailable"
COLLISION = "aggregate_collision"
OVER_LIMIT = "aggregate_over_limit"


@dataclass(frozen=True)
class Route:
    """A complete downstream destination; public names are never split."""

    public_name: str
    server_id: str
    downstream_tool_name: str


@dataclass(frozen=True)
class Snapshot:
    generation: int
    state: str
    tools: tuple[Mapping[str, Any], ...]
    routes: Mapping[str, Route]
    execution: Mapping[str, tuple[Route, Mapping[str, Any]]]
    result: Mapping[str, Any] | None

    @property
    def tool_count(self) -> int:
        return len(self.tools)


@dataclass(frozen=True)
class PreparedRoute:
    generation: int
    route: Route
    tool: Mapping[str, Any]
    policy_digest: str


class RouteMissing(LookupError):
    pass


class RouteChanged(RuntimeError):
    pass


class RouteDenied(PermissionError):
    pass


class Registry:
    """Own complete downstream catalogs and publish only finished snapshots."""

    def __init__(self, limits: LimitsConfig, policy: Policy, build_version: str) -> None:
        self._limits = limits
        self._policy = policy
        self._build_version = build_version
        self._catalogs: dict[str, tuple[Mapping[str, Any], ...]] = {}
        self._snapshot = Snapshot(0, STARTING, (), MappingProxyType({}), MappingProxyType({}), None)
        self._lock = asyncio.Lock()

    @property
    def snapshot(self) -> Snapshot:
        return self._snapshot

    async def publish(self, server_id: str, tools: list[Mapping[str, Any]]) -> Snapshot:
        owned = tuple(_frozen_json(tool) for tool in tools)
        async with self._lock:
            self._catalogs[server_id] = owned
            self._rebuild()
            return self._snapshot

    async def remove(self, server_id: str) -> Snapshot:
        async with self._lock:
            self._catalogs.pop(server_id, None)
            self._rebuild()
            return self._snapshot

    async def prepare(self, public_name: str) -> PreparedRoute:
        async with self._lock:
            if self._snapshot.state != READY or public_name not in self._snapshot.execution:
                raise RouteMissing(public_name)
            route, tool = self._snapshot.execution[public_name]
            return PreparedRoute(self._snapshot.generation, route, _frozen_json(tool), self._policy.digest)

    async def authorize(self, prepared: PreparedRoute, commit: Callable[[], Awaitable[None]]) -> None:
        """Hold publication while the caller durably records authorization."""

        async with self._lock:
            active = self._snapshot
            entry = active.execution.get(prepared.route.public_name)
            if active.state != READY or active.generation != prepared.generation or entry is None or entry[0] != prepared.route:
                raise RouteChanged(prepared.route.public_name)
            if not self._policy.allowed(prepared.route.public_name):
                raise RouteDenied(prepared.route.public_name)
            await commit()

    def _rebuild(self) -> None:
        generation = self._snapshot.generation + 1
        if not self._catalogs:
            self._snapshot = Snapshot(generation, UNAVAILABLE, (), MappingProxyType({}), MappingProxyType({}), None)
            return
        all_tools: dict[str, tuple[Mapping[str, Any], Mapping[str, Any], Route]] = {}
        for server_id, catalog in self._catalogs.items():
            for original in catalog:
                name = original.get("name") if isinstance(original, Mapping) else None
                if not isinstance(name, str) or not name:
                    self._snapshot = Snapshot(generation, UNAVAILABLE, (), MappingProxyType({}), MappingProxyType({}), None)
                    return
                public_name = f"{server_id}.{name}"
                if public_name in all_tools:
                    self._snapshot = Snapshot(generation, COLLISION, (), MappingProxyType({}), MappingProxyType({}), None)
                    return
                public = _thaw_json(original)
                public["name"] = public_name
                all_tools[public_name] = (_frozen_json(public), original, Route(public_name, server_id, name))
        names = sorted(name for name in all_tools if self._policy.allowed(name))
        if len(names) > self._limits.max_aggregate_tools:
            self._snapshot = Snapshot(generation, OVER_LIMIT, (), MappingProxyType({}), MappingProxyType({}), None)
            return
        tools = tuple(all_tools[name][0] for name in names)
        routes = MappingProxyType({name: all_tools[name][2] for name in names})
        execution = MappingProxyType({name: (entry[2], entry[1]) for name, entry in all_tools.items()})
        result = {
            "resultType": "complete",
            "_meta": {"io.modelcontextprotocol/serverInfo": {"name": "conduit", "version": self._build_version}},
            "ttlMs": 0,
            "cacheScope": "public",
            "tools": [_thaw_json(tool) for tool in tools],
        }
        if len(json.dumps(result, separators=(",", ":"), ensure_ascii=False).encode()) > self._limits.max_aggregate_response_bytes:
            self._snapshot = Snapshot(generation, OVER_LIMIT, (), MappingProxyType({}), MappingProxyType({}), None)
            return
        self._snapshot = Snapshot(generation, READY, tools, routes, execution, _frozen_json(result))


def _frozen_json(value: Mapping[str, Any]) -> Mapping[str, Any]:
    """Recursively detach JSON data before publishing a generation."""

    frozen = _freeze(value)
    if not isinstance(frozen, Mapping):  # defensive: tools and results are objects
        raise TypeError("JSON object expected")
    return frozen


def _thaw_json(value: Mapping[str, Any]) -> dict[str, Any]:
    thawed = _thaw(value)
    if not isinstance(thawed, dict):  # defensive: tools and results are objects
        raise TypeError("JSON object expected")
    return thawed


def _freeze(value: Any) -> Any:
    if isinstance(value, Mapping):
        return MappingProxyType({str(key): _freeze(item) for key, item in value.items()})
    if isinstance(value, list) or isinstance(value, tuple):
        return tuple(_freeze(item) for item in value)
    return value


def _thaw(value: Any) -> Any:
    if isinstance(value, Mapping):
        return {str(key): _thaw(item) for key, item in value.items()}
    if isinstance(value, tuple):
        return [_thaw(item) for item in value]
    return value
